Assets/BossRoom/Scripts/Shared/Infrastructure/DIScope.cs

```diff
@@ -0,0 +1,333 @@
+        private readonly Dictionary<Type, object> m_TypesToInstances = new Dictionary<Type, object>();
+        private bool m_Disposed;
+
+        private bool m_ScopeConstructionComplete;
+
+        public DIScope(DIScope parent = null)
+        {
+            m_Parent = parent;
+            BindInstanceAsSingle<IInstanceResolver, DIScope>(this);
+        }
+
+        public static DIScope RootScope
+        {
+            get
+            {
+                if (m_rootScope == null) m_rootScope = new DIScope();
+
+                return m_rootScope;
+            }
+        }
+
+        public void Dispose()
+        {
+            if (!m_Disposed)
+            {
+                m_TypesToInstances.Clear();
+                m_DisposableGroup.Dispose();
+                m_Disposed = true;
+            }
+        }
+
+        public T Resolve<T>()
+            where T : class
+        {
+            if (!m_ScopeConstructionComplete)
+                throw new ScopeNotFinalizedException(
+                    $"Trying to Resolve type {typeof(T)}, but the DISCope is not yet finalized! You should call FinalizeScopeConstruction before any of the Resolve calls.");
+            //if we have this type as lazy-bound instance - we are going to instantiate it now
+            if (m_LazyBindDescriptors.TryGetValue(typeof(T), out var lazyBindDescriptor))
+            {
+                var instance = (T) InstantiateLazyBoundObject(lazyBindDescriptor);
+                m_LazyBindDescriptors.Remove(typeof(T));
+                return instance;
+            }
+
+            if (!m_TypesToInstances.TryGetValue(typeof(T), out var value))
+            {
+                if (m_Parent != null) return m_Parent.Resolve<T>();
+
+                throw new NoInstanceToInjectException($"Injection of type {typeof(T)} failed.");
+            }
+
+            return (T) value;
+        }
+
+        public void InjectIn(object obj)
+        {
+            if (CachedReflectionUtility.TryGetInjectableMethod(obj.GetType(), out var injectionMethod))
+            {
+                var parameters = CachedReflectionUtility.GetMethodParameters(injectionMethod);
+
+                var paramColleciton = new object[parameters.Length];
+
+                for (var i = 0; i < parameters.Length; i++)
+                {
+                    var parameter = parameters[i];
+
+                    var genericResolveMethod = CachedReflectionUtility.GetTypedResolveMethod(parameter.ParameterType);
+                    var resolved = genericResolveMethod.Invoke(this, null);
+                    paramColleciton[i] = resolved;
+                }
+
+                injectionMethod.Invoke(obj, paramColleciton);
+            }
+        }
+
+        public void InjectIn(GameObject go)
+        {
+            var components = go.GetComponentsInChildren<Component>();
+
+            foreach (var component in components) InjectIn(component);
+        }
+
+        ~DIScope()
```

Comment:
general formatting, usually you like having your destructor close to your constructor :)

Reply:
Ah crap, that's some automatic code cleanup that I did - that horrible foreach is also caused by the same... Or is it our actual codestyle? Anyhow - you're right - all of the creation/destruction logic belongs together.